pkg/deployment/context_impl.go

```diff
@@ -63,20 +64,44 @@ func (d *Deployment) GetNamespace() string {
 	return d.apiObject.GetNamespace()
 }
 
+// GetPhase returns the current phase of the deployment
+func (d *Deployment) GetPhase() api.DeploymentPhase {
+	return d.status.last.Phase
+}
+
 // GetSpec returns the current specification
 func (d *Deployment) GetSpec() api.DeploymentSpec {
 	return d.apiObject.Spec
 }
 
 // GetStatus returns the current status of the deployment
-func (d *Deployment) GetStatus() api.DeploymentStatus {
-	return d.status
+// together with the current version of that status.
+func (d *Deployment) GetStatus() (api.DeploymentStatus, int32) {
+	d.status.mutex.Lock()
+	defer d.status.mutex.Unlock()
+
+	version := d.status.version
+	return *d.status.last.DeepCopy(), version
 }
 
 // UpdateStatus replaces the status of the deployment with the given status and
 // updates the resources in k8s.
-func (d *Deployment) UpdateStatus(status api.DeploymentStatus, force ...bool) error {
-	d.status = status
+// If the given last version does not match the actual last version of the status object,
+// an error is returned.
+func (d *Deployment) UpdateStatus(status api.DeploymentStatus, lastVersion int32, force ...bool) error {
+	d.status.mutex.Lock()
+	defer d.status.mutex.Unlock()
+
+	if d.status.version != lastVersion {
+		// Status is obsolete
+		d.deps.Log.Error().
+			Int32("expected-version", lastVersion).
+			Int32("actual-version", d.status.version).
+			Msg("UpdateStatus version conflict error.")
+		return maskAny(fmt.Errorf("Status conflict error. Expected version %d, got %d", lastVersion, d.status.version))
+	}
+	d.status.version++
+	d.status.last = *status.DeepCopy()
 	if err := d.updateCRStatus(force...); err != nil {
 		return maskAny(err)
```

Comment:
If there is a retry higher up, it MUST call `GetStatus()` first to get the latest status and version.
So while a network error will cause a state diff between in-memory & persistent in kube-apiserver, this will not affect future updates.

Reply:
ok